tests/throttle.cpp

```diff
@@ -26,32 +26,69 @@
 
 #include "../plugins/RocketsPlugin/Throttle.h"
 
+struct TestTimer
+{
+    void start() { timer.start(); }
+    void onCalled()
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        timer.stop();
+        called = true;
+        condition.notify_one();
+    }
+
+    void wait()
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        while (!called)
+            condition.wait(lock);
+    }
+
+    void wait_for(const int64_t waitTime)
+    {
+        std::unique_lock<std::mutex> lock(mutex);
+        condition.wait_for(lock, std::chrono::milliseconds(waitTime));
+    }
+
+    auto milliseconds() const { return timer.milliseconds(); }
+    auto wasCalled() const { return called; }
+private:
+    bool called = false;
+    std::mutex mutex;
+    std::condition_variable condition;
+    brayns::Timer timer;
+};
+
 BOOST_AUTO_TEST_CASE(timeout_not_cleared)
 {
     brayns::Timeout timeout;
-    std::atomic_bool called{false};
-    timeout.set([&called] { called = true; }, 1);
-    std::this_thread::sleep_for(std::chrono::milliseconds(2));
-    BOOST_CHECK(called);
+
+    const int64_t waitTime = 5;
+    TestTimer timer;
+    timer.start();
+    timeout.set([&] { timer.onCalled(); }, waitTime);
+    timer.wait();
+
+    BOOST_CHECK_GE(timer.milliseconds(), waitTime);
+    BOOST_CHECK(timer.wasCalled());
 }
 
 BOOST_AUTO_TEST_CASE(timeout_with_clear)
 {
-    brayns::Timer timer;
-    timer.start();
-
     brayns::Timeout timeout;
 
-    std::atomic_bool called{false};
-    timeout.set([&called] { called = true; }, 500);
+    const int64_t waitTime = 10;
+    TestTimer timer;
+    timer.start();
+    timeout.set([&] { timer.onCalled(); }, waitTime);
     timeout.clear();
-    BOOST_CHECK(!called);
+    timer.wait_for(waitTime);
 
-    timer.stop();
-    BOOST_CHECK_LE(timer.milliseconds(), 500);
+    BOOST_CHECK_LT(timer.milliseconds(), waitTime);
 
-    std::this_thread::sleep_for(std::chrono::milliseconds(2));
-    BOOST_CHECK(!called);
+    std::this_thread::sleep_for(
```

Comment:
Once the code gets here, it should normally be safe to BOOST_CHECK(!timer.wasCalled()); right? So is this extra wait here to cover the case of spurrious wakeup in TestTimer::wait_for? If that's the case, I noticed that there is an overload of condition.wait_for() that takes a predicate (here that would be []{!called}) that takes care of removing the spurrious wakeups. It could make the timer.wait_for() call predicatable and simplify the testing.

Reply:
The sleep does not make sense anymore, good catch. The spurious wakeup could be an issue, I will look at this predicate for wait_for, thanks!